applications/asset_tracker/src/cloud_codec/cloud_codec.c

```diff
@@ -0,0 +1,385 @@
+			CMD_NEW_TYPE(read, CLOUD_CMD_READ),
+			CMD_NEW_TYPE(read_all, CLOUD_CMD_READ))
+		))
+	),
+	CMD_NEW_RECIPIENT(ui, CLOUD_RCPT_UI, CMD_ARRAY(
+		CMD_NEW_CHAN(pin, CLOUD_CHANNEL_PIN, CMD_ARRAY(
+			CMD_NEW_TYPE(read, CLOUD_CMD_READ),
+			CMD_NEW_TYPE(read_all, CLOUD_CMD_READ))
+		),
+		CMD_NEW_CHAN(led, CLOUD_CHANNEL_RGB_LED, CMD_ARRAY(
+			CMD_NEW_TYPE(read, CLOUD_CMD_READ),
+			CMD_NEW_TYPE(read_all, CLOUD_CMD_READ))
+		),
+		CMD_NEW_CHAN(buzzer, CLOUD_CHANNEL_BUZZER, CMD_ARRAY(
+			CMD_NEW_TYPE(read, CLOUD_CMD_READ),
+			CMD_NEW_TYPE(read_all, CLOUD_CMD_READ))
+		))
+	))
+);
+
+static const char *const channel_type_str[] = {
+	[CLOUD_CHANNEL_GPS] = "GPS",
+	[CLOUD_CHANNEL_FLIP] = "FLIP",
+	[CLOUD_CHANNEL_BUTTON] = "BUTTON",
+	[CLOUD_CHANNEL_TEMP] = "TEMP",
+	[CLOUD_CHANNEL_HUMID] = "HUMID",
+	[CLOUD_CHANNEL_AIR_PRESS] = "AIR_PRESS",
+	[CLOUD_CHANNEL_AIR_QUAL] = "AIR_QUAL",
+	[CLOUD_CHANNEL_LTE_LINK_RSRP] = "RSRP",
+	[CLOUD_CHANNEL_DEVICE_INFO] = "DEVICE",
+};
+
+static cloud_cmd_cb_t cloud_command_cb;
+struct cloud_command cmd_parsed;
+
+/* --- A few wrappers for cJSON APIs --- */
+
+static int json_add_obj(cJSON *parent, const char *str, cJSON *item)
+{
+	cJSON_AddItemToObject(parent, str, item);
+
+	return 0;
+}
+
+static int json_add_num(cJSON *parent, const char *str, double num)
+{
+	cJSON *json_num;
+
+	json_num = cJSON_CreateNumber(num);
+	if (json_num == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_num);
+}
+
+static int json_add_str(cJSON *parent, const char *str, const char *item)
+{
+	cJSON *json_str;
+
+	json_str = cJSON_CreateString(item);
+	if (json_str == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_str);
+}
+
+static int json_add_null(cJSON *parent, const char *str)
+{
+	cJSON *json_null;
+
+	json_null = cJSON_CreateNull();
+	if (json_null == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_null);
+}
+
+static cJSON *json_object_decode(cJSON *obj, const char *str)
+{
+	return obj ? cJSON_GetObjectItem(obj, str) : NULL;
+}
+
+int cloud_encode_data(const struct cloud_channel_data *channel,
+			struct cloud_data *output)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(channel != NULL);
+	__ASSERT_NO_MSG(channel->data.buf != NULL);
+	__ASSERT_NO_MSG(channel->data.len != 0);
+	__ASSERT_NO_MSG(output != NULL);
+
+	cJSON *root_obj = cJSON_CreateObject();
+
+	if (root_obj == NULL) {
+		return -ENOMEM;
+	}
+
+	ret = json_add_str(root_obj, "appId", channel_type_str[channel->type]);
+	ret += json_add_str(root_obj, "data", channel->data.buf);
+	ret += json_add_str(root_obj, "messageType", "DATA");
+
+	if (ret != 0) {
+		cJSON_Delete(root_obj);
+		return -ENOMEM;
+	}
+
+	char *buffer;
+
+	buffer = cJSON_PrintUnformatted(root_obj);
+	cJSON_Delete(root_obj);
+
+	output->buf = buffer;
+	output->len = strlen(buffer);
+
+	return 0;
+}
+
+static int cloud_search_cmd(cJSON *group_obj, enum cloud_cmd_group group)
+{
+	__ASSERT_NO_MSG(group_obj != 0);
```

Comment:
Variable declaration should come on top of the block

Reply:
Fixed